File: backend/app/v1/utils/stage.py
import uuid
from typing import Any

def get_stage_required_inputs(stage_config: dict | None, template_name: str | None) -> list[str]:
    """
    Resolves the required inputs for a stage.
    Supports a dynamic 'required_inputs' list.
    If not specified, falls back to default values based on the template name
    for backward compatibility.
    """
    if isinstance(stage_config, dict) and "required_inputs" in stage_config:
        inputs = stage_config["required_inputs"]
        # Only return the inputs if it is a list and has at least one element.
        # Otherwise, fall back to default values.
        if isinstance(inputs, list) and len(inputs) > 0:
            return [str(i) for i in inputs]
            
    # Fallback rules based on template name
    if template_name == "Technical Practical Round":
        return ["github", "question"]
    elif template_name == "Resume Screening":
        return ["resume"]
    else:
        return ["transcript"]



def get_question_round_filter(job_stage_config_cls, stage_template_cls):
    """
    Returns the SQLAlchemy filter to check if a stage requires 'question'.
    Falls back to template name == 'Technical Practical Round' if required_inputs is absent.
    """
    from sqlalchemy import or_, and_, func
    
    config_has_question = job_stage_config_cls.config["required_inputs"].contains(["question"])
    template_has_question = stage_template_cls.default_config["required_inputs"].contains(["question"])
    
    config_missing_or_empty = or_(
        job_stage_config_cls.config.is_(None), 
        ~job_stage_config_cls.config.has_key("required_inputs"),
        func.jsonb_array_length(job_stage_config_cls.config["required_inputs"]) == 0
    )
    
    template_missing_or_empty = or_(
        stage_template_cls.default_config.is_(None), 
        ~stage_template_cls.default_config.has_key("required_inputs"),
        func.jsonb_array_length(stage_template_cls.default_config["required_inputs"]) == 0
    )
    
    required_inputs_absent_and_name_matches = and_(
        config_missing_or_empty,
        template_missing_or_empty,
        stage_template_cls.name == "Technical Practical Round"
    )
    
    return or_(
        config_has_question,
        and_(
            config_missing_or_empty,
            template_has_question
        ),
        required_inputs_absent_and_name_matches
    )

File: frontend/src/utils/stage.ts
/**
 * @module stage
 * Utility functions for evaluating stage configurations and template types in the hiring process workflow.
 */

/**
 * Determines if a stage is a question round (which typically requires task papers assignment).
 * Checks if the stage configuration explicitly includes "question" in the `required_inputs`,
 * or falls back to matching the template name `"Technical Practical Round"`.
 *
 * @param stageConfig - The configuration object for the hiring stage, or null/undefined.
 * @returns `true` if the stage is identified as a question round, otherwise `false`.
 */
export function isQuestionStage(stageConfig?: { config?: any; template?: { name: string; config?: any } } | null): boolean {
  if (!stageConfig) return false;
  const config = stageConfig.config || stageConfig.template?.config;
  const templateName = stageConfig.template?.name;
  
  if (config && Array.isArray(config.required_inputs) && config.required_inputs.length > 0) {
    return config.required_inputs.includes("question");
  }
  
  return templateName === "Technical Practical Round";
}

/**
 * Determines if a stage is a transcript evaluation round.
 * Checks if the stage configuration explicitly includes "transcript" in the `required_inputs`,
 * or falls back to checking that the template name is neither `"Technical Practical Round"` nor `"Resume Screening"`.
 *
 * @param stageConfig - The configuration object for the hiring stage, or null/undefined.
 * @returns `true` if the stage is identified as a transcript round, otherwise `false`.
 */
export function isTranscriptStage(stageConfig?: { config?: any; template?: { name: string; config?: any } } | null): boolean {
  if (!stageConfig) return false;
  const config = stageConfig.config || stageConfig.template?.config;
  const templateName = stageConfig.template?.name;
  
  if (config && Array.isArray(config.required_inputs) && config.required_inputs.length > 0) {
    return config.required_inputs.includes("transcript");
  }
  
  return templateName !== "Technical Practical Round" && templateName !== "Resume Screening";
}
